src/broker/cas_cgw.c

```diff
@@ -2438,3 +2455,277 @@ cgw_utype_to_string (int type)
+    {
+      is_rewrite_error = true;
+      goto REWRITE_ERROR;
+    }
+
+  new_inline_view_len += REWRITE_SELECT_FROM_LEN;
+  new_inline_view_len += strlen (" AS ") * num_cols;
+  new_inline_view_len += inline_view_len;
+  new_inline_view_len += cublink_col_list_len;
+
+  for (int i = 1; i <= num_cols; i++)
+    {
+      err_code = SQLColAttribute (hstmt, i, SQL_DESC_NAME, col_name, sizeof (col_name), &col_name_length, NULL);
+      new_inline_view_len += col_name_length;
+    }
+
+  new_inline_view = (char *) MALLOC (new_inline_view_len + 1);
+  strcpy (new_inline_view, "(SELECT ");
+
+  for (int i = 1; i <= num_cols; i++)
+    {
+      err_code = SQLColAttribute (hstmt, i, SQL_DESC_NAME, col_name, sizeof (col_name), &col_name_length, NULL);
+
+      if (err_code < 0)
+	{
+	  is_rewrite_error = true;
+	  goto REWRITE_ERROR;
+	}
+
+      if (cols_list[i - 1] == NULL || strcmp (cols_list[i - 1], "") == 0)
+	{
+	  is_rewrite_error = true;
+	  goto REWRITE_ERROR;
+	}
+
+      strcat (new_inline_view, col_name);
+      strcat (new_inline_view, " AS ");
+      strcat (new_inline_view, cols_list[i - 1]);
+
+      if (i <= num_cols - 1)
+	{
+	  strcat (new_inline_view, ", ");
+	}
+      else
+	{
+	  strcat (new_inline_view, " FROM ");
+	  strcat (new_inline_view, inline_view);
+	  strcat (new_inline_view, ")");
+	}
+    }
+
+  rewrite_query = (char *) MALLOC (REWRITE_SELECT_FROM_LEN + new_inline_view_len + where_len);
+  memset (rewrite_query, 0x0, REWRITE_SELECT_FROM_LEN + new_inline_view_len + where_len);
+
+  strcat (rewrite_query, select_from);
+  strcat (rewrite_query, " ");
+  strcat (rewrite_query, new_inline_view);
+  strcat (rewrite_query, " ");
+  strcat (rewrite_query, where);
+
+REWRITE_ERROR:
+  SQLFreeHandle (SQL_HANDLE_STMT, hstmt);
+
+  if (select_from != NULL)
+    {
+      FREE_MEM (select_from);
+    }
+
+  if (inline_view != NULL)
+    {
+      FREE_MEM (inline_view);
+    }
+
+  if (cublink != NULL)
+    {
+      FREE_MEM (cublink);
+    }
+
+  if (cols_type != NULL)
+    {
+      FREE_MEM (cols_type);
+    }
+
+  if (cols_list != NULL)
+    {
+      cgw_free_string_array (cols_list);
+    }
+
+  if (where != NULL)
+    {
+      FREE_MEM (where);
+    }
+
+  if (rewrite_query != NULL && is_rewrite_error == TRUE)
+    {
+      FREE_MEM (rewrite_query);
+      rewrite_query = NULL;
+    }
+
+  return rewrite_query;
+}
+
+static void
+cgw_eliminate (char *str, char ch)
+{
+  size_t len = strlen (str) + 1;
+  for (; *str != '\0'; str++, len--)
+    {
+      if (*str == ch || *str == ' ')
+	{
+	  strncpy (str, str + 1, len);
+	  str--;
+	}
+    }
+}
+
+static void
+cgw_free_string_array (char **array)
+{
+  int i;
+
+  for (i = 0; array[i] != NULL; i++)
+    {
+      FREE_MEM (array[i]);
+    }
+  FREE_MEM (array);
+}
+
+static char **
+cgw_split_string (const char *str, const char *delim)
+{
+  char *t, *o;
+  char *v;
+  char **r = NULL;
+  int count = 1;
+
+  if (str == NULL)
+    {
+      return NULL;
+    }
+
+  o = strdup (str);
```

Comment:
str is already not NULL, this check is not necessary also.

Reply:
This is the defense code.